subsys/net/ip/tcp.c

```diff
@@ -2417,6 +2417,19 @@ static enum tcp_state tcp_enter_time_wait(struct tcp *conn)
 	return TCP_TIME_WAIT;
 }
 
+static void tcp_fin_wait_2_timeout(struct k_work *work)
```

Comment:
I'm not actually sure if we need a specific timeout logic for FIN_WAIT_2 exclusively. The fin timer should always be set when we enter FIN_WAIT_1 state (which is a prerequisite for FIN_WAIT_2), and should not be reset until the final FIN/ACK reply arrives or timeout occurs. It seems to me, that we're simply missing setting up the timer here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L3053
At least it helps for the case when we're stuck in FIN_WAIT_2 with burp tool (we were stuck as the timer was not set when entered FIN_WAIT_1)